cmake/modules/extensions.cmake

```diff
@@ -1283,13 +1283,67 @@ endfunction(zephyr_linker_sources)
 
 
 # Helper function for CONFIG_CODE_DATA_RELOCATION
-# Call this function with 2 arguments file and then memory location.
-# One optional [NOCOPY] flag can be used.
-function(zephyr_code_relocate file location)
+# This function may either be invoked with a list of files, or a library
+# name to relocate.
+#
+# The FILES directive will relocate a list of files (wildcards supported)
+# This directive will relocate file1. and file2.c to SRAM:
+# zephyr_code_relocate(FILES file1.c file2.c LOCATION SRAM)
+# Note, files can also be passed as a comma separated list to support using
+# cmake generator arguments
+#
+# The LIBRARY directive will relocate a library
+# This directive will relocate the target my_lib to SRAM:
+# zephyr_code_relocate(LIBRARY my_lib SRAM)
+#
+# One optional [NOCOPY] flag can be used, to indicate that the file data does
+# not need to be copied at boot time (For example, for flash XIP).
+function(zephyr_code_relocate)
   set(options NOCOPY)
-  cmake_parse_arguments(CODE_REL "${options}" "" "" ${ARGN})
-  if(NOT IS_ABSOLUTE ${file})
-    set(file ${CMAKE_CURRENT_SOURCE_DIR}/${file})
+  set(single_args LIBRARY LOCATION)
+  set(multi_args FILES)
+  cmake_parse_arguments(CODE_REL "${options}" "${single_args}"
+    "${multi_args}" ${ARGN})
+  # Argument validation
+  if(CODE_REL_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_code_relocate(${ARGV0} ...) "
+      "given unknown arguments: ${CODE_REL_UNPARSED_ARGUMENTS}")
+  endif()
+  if((NOT CODE_REL_FILES) AND (NOT CODE_REL_LIBRARY))
+    message(FATAL_ERROR
+      "zephyr_code_relocate() requires either FILES or LIBRARY be provided")
+  endif()
+  if(CODE_REL_FILES AND CODE_REL_LIBRARY)
+    message(FATAL_ERROR "zephyr_code_relocate() only accepts "
+      "one argument between FILES and LIBRARY")
+  endif()
+  if(NOT CODE_REL_LOCATION)
+    message(FATAL_ERROR "zephyr_code_relocate() requires a LOCATION argument")
+  endif()
+  if(CODE_REL_LIBRARY)
+    # Use cmake generator expression to convert library to file list
+    set(genex_src_dir "$<TARGET_PROPERTY:${CODE_REL_LIBRARY},SOURCE_DIR>")
+    set(genex_src_list "$<TARGET_PROPERTY:${CODE_REL_LIBRARY},SOURCES>")
+    set(file_list
+      "${genex_src_dir}/$<JOIN:${genex_src_list},$<COMMA>${genex_src_dir}/>")
+  else()
+    # Check if CODE_REL_FILES is a generator expression, if so leave it
+    # untouched.
+    string(GENEX_STRIP "${CODE_REL_FILES}" no_genex)
+    if(CODE_REL_FILES STREQUAL no_genex)
+      # no generator expression in CODE_REL_FILES, check if list of files
+      # is absolute
+      foreach(file ${CODE_REL_FILES})
+        if(NOT IS_ABSOLUTE ${file})
+          set(file ${CMAKE_CURRENT_SOURCE_DIR}/${file})
+        endif()
+        list(APPEND file_names ${file})
+      endforeach()
+      set(file_list "$<JOIN:${file_names},$<COMMA>>")
+    else()
+      # Generator expression is present in file list. Leave the list untouched.
+      set(file_list ${CODE_REL_FILES})
```

Comment:
I think the substitution method is probably best, with some documentation provided to the user. It is unfortunate, but if we don't perform some type of modification on the user input then the other alternative I can see be to change the genex directive in [target_relocation.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/ae9de202220ce51e1ed61157f73d3800f84c6b06/cmake/linker/ld/target_relocation.cmake#L21) to something like the following:

```
-i \"$<JOIN:$<TARGET_PROPERTY:code_data_relocation_target,COMPILE_DEFINITIONS>,$<COMMA>>\"
```
which would require the genex passed to `zephyr_library_relocate` to be modified, otherwise this `JOIN` directive would change it as well. So either way, we are left needing to change user input.

If we wanted to place the responsibility on the caller, we could update the `gen_relocate_app.py` script to print a warning when it detects a relocation directive without `:` present, which would be the failure mode of the script when a genex expression was passed without `,` subsituted

Reply:
@tejlmand What are your thoughts here? I'd like to move this PR forwards, as #50792 depends on a feature like this being available. Once we settle on a method here I can make other requested updates